### PR TITLE
data cleaning for institution list used for user map #4970

### LINKED ISSUE #4970: Prune invalid and duplicate inputs in institution list
This is a follow up issue for #4872 as the institution list data seem to contain quite a number of invalid and duplicate inputs.

I will implement a pruning function and prune the list before actually using the list.

### PR DESCRIPTION
**Pull request by paradite**
fixes #4970 

I have implemented data parsing and cleaning logic for the raw institution list.
For institutions that are missing the country, I've added them manually.
For institutions that are missing the country but already have duplicates, I've removed them manually.

### CODE REVIEW

**Comment by damithc** [comment]
Just cleaning the list manually is not enough? I wasn't expecting functions to clean the list. Once the list is clean, we'll ensure future additions don't make it dirty again.

**Comment by paradite** [comment]
@damithc if you are adding items manually into the list in the future, I remove the functions replace the data with the cleaned data, but it will be in a different, unified format, with each entry having two elements, `[name, country]`.
if we use the functions, it will handle the updated data automatically, with slight parsing of the raw data into valid JSON structure.

**Comment by damithc** [comment]
Yes, it will be done manually. Yes, a different format is ok as long as it is not difficult to update manually.
Perhaps you can sort it by country, as we discussed?

**Comment by paradite** [comment]
@damithc updated. I've also changed the cleaned data into a JSON file since it is just data without any logic.

**Comment by wkurniawan07** [comment]
Can you sort alphabetically within a country itself? E.g:
```javascript
// other countries, e.g Scotland
"Nanyang Technological University", "Singapore"
"National University of Singapore", "Singapore"
"Singapore Management University", "Singapore"
// other countries, e.g South Africa
```
There are still a handful of duplications and alphabetical sorting might help to catch.

**Comment by paradite** [comment]
@wkurniawan07 done

File: src/main/webapp/js/userMapData.json [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+["Myeongju Elementary School","South Korea"],
+["Centro Didáctico","Spain"],
+["IE Business School","Spain"],
+["IESE Business School","Spain"],
+["Mondragon University","Spain"],
+["Santa María del Pilar","Spain"],
+["Universidad Complutense","Spain"],
+["Universidad Complutense de Madrid","Spain"],
+["Universidad de Navarra","Spain"],
+["Universidad del Pa? Vasco","Spain"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Correct this one.

File: src/main/webapp/js/userMapData.json [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+["Ngee Ann Polytechnic","Singapore"],
+["Punggol Green Primary School","Singapore"],
+["Raffles College of Higher Education","Singapore"],
+["Raffles Junior College","Singapore"],
+["Sengkang Secondary School","Singapore"],
+["SHATEC Pte Ltd","Singapore"],
+["Silicon Straits","Singapore"],
+["Singapore Chinese Girls' School","Singapore"],
+["Singapore Institute of Technology","Singapore"],
+["Singapore Management University","Singapore"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: On on slightly unrelated note, how do we deal with the problem of Singapore being too small to show up in the map?

File: src/main/webapp/js/userMapData.json [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+["The Petroleum Institute","UAE"],
+["Cardiff University","UK"],
+["Cranfield University","UK"],
+["Edinburgh Napier University","UK"],
+["Emmanuel College","UK"],
+["Home Learning College","UK"],
+["Imperial College London","UK"],
+["Leeds Trinity University","UK"],
+["London School of Economics","UK"],
+["LSE","UK"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: This two are the same

File: src/main/webapp/js/userMapData.json [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+["The University of Manchester","UK"],
+["University College London","UK"],
+["University of Bath","UK"],
+["University of Bradford","UK"],
+["University of Bristol","UK"],
+["University of Central Lancashire","UK"],
+["University of Chichester","UK"],
+["University of Hertfordshire","UK"],
+["University of Huddersfield","UK"],
+["University of Manchester","UK"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Duplicated above

File: src/main/webapp/js/userMapData.json [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+["Columbia University","USA"],
+["Connally High School","USA"],
+["Cornell University","USA"],
+["Creighton University","USA"],
+["Cumberland Valley High School","USA"],
+["Cuyahoga Community College","USA"],
+["Cuyamaca College","USA"],
+["Cypress College","USA"],
+["Cypress Heights Academy","USA"],
+["D&#39;Youville College","USA"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Dirty data

File: src/main/webapp/js/userMapData.json [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+["Georgia Institute of Technology","USA"],
+["Georgia State University","USA"],
+["Glendale Community College","USA"],
+["GLOBALFOUNDRIES","USA"],
+["Goldenwest College","USA"],
+["Goreville High School","USA"],
+["Grand Valley State University","USA"],
+["Greensboro Day School","USA"],
+["Griffin  Middle School","USA"],
+["Griffin Middle School","USA"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Duplicates

**Comment by damithc** [comment]
The list is not clean enough yet. I've caught a few but I'm sure there are others. Do a more thorough check. Manual eyeballing required.

File: src/main/webapp/js/userMapData.json (hunk 1 above)
Comment (paradite) [inline src/main/webapp/js/userMapData.json]: It is like this in the raw file `2016-02-institutues`:
line 636
[Ithaca College, USA]
[Universidad del Pa? Vasco, Spain]
[Leeds Trinity University, UK]

Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Yes, the original data is dirty due to various reasons. Clean as much as you can.

Comment (paradite) [inline src/main/webapp/js/userMapData.json]: So how to deal with entries like this?

File: src/main/webapp/js/userMapData.json (hunk 5 above)
Comment (paradite) [inline src/main/webapp/js/userMapData.json]: this is also from the original file, how to deal with this?

File: src/main/webapp/js/userMapData.json (hunk 2 above)
Comment (paradite) [inline src/main/webapp/js/userMapData.json]: I can open another PR and implement the zoom feature. I've looked it up and it is not very hard.

File: src/main/webapp/js/userMapData.json (hunk 1 above)
Comment (paradite) [inline src/main/webapp/js/userMapData.json]: never mind I googled it.

**Comment by paradite** [comment]
@damithc I've tried my best in removing the duplicates, let me know if it's okay.

File: src/main/webapp/js/userMapData.json [hunk 7]
```diff
@@ -0,0 +1,793 @@
+[
+["Southworks","Argentina"],
+["University of Aruba","Aruba"],
+["Australian National University","Australia"],
+["Bond University","Australia"],
+["Curtin University","Australia"],
+["Deakin University","Australia"],
+["Edith Cowan University","Australia"],
+["Jacquelyn Humphrey","Australia"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Remove this one

File: src/main/webapp/js/userMapData.json [hunk 8]
```diff
@@ -0,0 +1,793 @@
...
+["Monash University","Australia"],
+["Photography Studies College","Australia"],
+["RMIT University","Australia"],
+["St Catherine's Catholic College","Australia"],
+["St Ives High School","Australia"],
+["St. Scholastica's College","Australia"],
+["Star of the Sea Catholic College","Australia"],
+["Swinburne Online","Australia"],
+["Swinburne University","Australia"],
+["Swinburne University of Technology","Australia"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Check if these three are the same.

File: src/main/webapp/js/userMapData.json [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+["Tabor Adelaide","Australia"],
+["The University of Melbourne","Australia"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Remove 'The'?

File: src/main/webapp/js/userMapData.json [hunk 10]
```diff
@@ -0,0 +1,793 @@
...
+["Toledo Prudente Centro Universitário","Brazil"],
+["UFRN","Brazil"],
+["Universidade Católica de Brasília","Brazil"],
+["Universidade de São Paulo","Brazil"],
+["Sekolah Menengah Sayyidina Ali","Brunei Darussalam"],
+["Trakia University","Bulgaria"],
+["Algonquin College","Canada"],
+["Aussie Pet Mobile Groomer Training","Canada"],
+["BCIT","Canada"],
+["British Columbia Institute of Technology","Canada"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: These two are the same

File: src/main/webapp/js/userMapData.json [hunk 11]
```diff
@@ -0,0 +1,793 @@
...
+["Universiti Putra Malaysia","Malaysia"],
+["Universiti Teknologi Malaysia","Malaysia"],
+["University College of Technology Sarawak","Malaysia"],
+["University Kuala Lumpur","Malaysia"],
+["University of Nottingham Malaysia Campus","Malaysia"],
+["Escuela Bancaria y Comercial","Mexico"],
+["ITESM","Mexico"],
+["ITESO","Mexico"],
+["Peterson Schools","Mexico"],
+["Tec de Monterrey","Mexico"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: This is same as ITESM

File: src/main/webapp/js/userMapData.json [hunk 12]
```diff
@@ -0,0 +1,793 @@
...
+["Tecnologico de Monterrey","Mexico"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: This too is same as ITESM

File: src/main/webapp/js/userMapData.json [hunk 13]
```diff
@@ -0,0 +1,793 @@
...
+["University of Manila","Philippines"],
+["University of the Cordilleras","Philippines"],
+["University of the Philippines","Philippines"],
+["Xavier School","Philippines"],
+["Adam Mickiewicz University","Poland"],
+["Universidade do Minho","Portugal"],
+["Universidad del Este","Puerto Rico"],
+["Carnegie Mellon University","Qatar"],
+["Qatar University","Qatar"],
+["SUNY Korea","Republic of Korea"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: This should be under Korea. And I think it is a duplicate.

File: src/main/webapp/js/userMapData.json (hunk 2 above)
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Sure. Let's do that.

File: src/main/webapp/js/userMapData.json [hunk 14]
```diff
@@ -0,0 +1,793 @@
...
+["Yale-NUS","Singapore"],
+["SC Nova Gorica","Slovenia"],
+["Univerza v Ljubljani","Slovenia"],
+["Kingsmead College","South Africa"],
+["Stellenbosch University","South Africa"],
+["University of Pretoria","South Africa"],
+["University of the Western Cape","South Africa"],
+["Harvard University","South Korea"],
+["Korea University Business School","South Korea"],
+["Myeongju Elementary School","South Korea"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: This three should be under Korea. Or the other Korea ones should be moved here.

File: src/main/webapp/js/userMapData.json [hunk 15]
```diff
@@ -0,0 +1,793 @@
...
+["Northern Kentucky University","USA"],
+["Northern Michigan University","USA"],
+["Northern Nash High School","USA"],
+["Northwestern Michigan College","USA"],
+["Northwestern University","USA"],
+["Northwood University","USA"],
+["Oakton Community College","USA"],
+["Ogeechee Technical College","USA"],
+["Ohio State University","USA"],
+["Ohio University","USA"],
```
Comment (damithc) [inline src/main/webapp/js/userMapData.json]: Check if these two are different universities

Comment (paradite) [inline src/main/webapp/js/userMapData.json]: they are different

File: src/main/webapp/js/userMapData.json (hunk 14 above)
Comment (paradite) [inline src/main/webapp/js/userMapData.json]: Moved the other over here

**Comment by paradite** [comment]
@damithc done

**Comment by damithc** [comment]
OK. Let's wait for reviewer clearance.

**Comment by paradite** [comment]
@wkurniawan07 please review. thanks.

File: src/main/webapp/js/userMapData.json [hunk 16]
```diff
@@ -0,0 +1,787 @@
...
+["ITESM","Mexico"],
+["ITESO","Mexico"],
+["Peterson Schools","Mexico"],
+["Universidad Autónoma de Baja California","Mexico"],
+["Energized.Org","Netherland"],
+["HAN University of Applied Sciences","Netherland"],
+["Hogeschool van Amsterdam","Netherland"],
+["Inholland University of Applied Sciences","Netherland"],
+["Kaj Munk College","Netherland"],
+["VU University","Netherland"],
```
Comment (wkurniawan07) [inline src/main/webapp/js/userMapData.json]: I don't think we need another 'Netherland'. Also, I believe the Hogeschool up there is the same as the one down.

File: src/main/webapp/js/userMapData.json [hunk 17]
```diff
@@ -0,0 +1,787 @@
...
+["University of the Pacific","USA"],
+["University of Toledo","USA"],
+["University of Virginia","USA"],
+["University of Washington","USA"],
+["University of Wisconsin","USA"],
+["University of Wisconsin Oshkosh","USA"],
+["University of Wisconsin School of Business","USA"],
+["University of Wisconsin-Madison","USA"],
+["University of Wisconsin-Platteville","USA"],
+["Unversity of San Diego","USA"],
```
Comment (wkurniawan07) [inline src/main/webapp/js/userMapData.json]: This is a duplicate, there's a non-typo version up there :p

File: src/main/webapp/js/userMap.js [hunk 18]
```diff
@@ -1,18 +1,17 @@
-document.addEventListener('DOMContentLoaded', function(event) {
+function handleData(err, userData) {
     // based on example from https://github.com/markmarkoh/datamaps/blob/master/src/examples/highmaps_world.html
     // Country code: https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3
-
+    if (err) {
+        console.log('error in getting data');
```
Comment (wkurniawan07) [inline src/main/webapp/js/userMap.js]: This is hardly helpful. Should show the error to the user (and ask them to reload the page) via status message or alike.

**Comment by wkurniawan07** [comment]
Few cleanups and one last major change requested.

File: src/main/webapp/js/userMap.js (hunk 18 above)
Comment (paradite) [inline src/main/webapp/js/userMap.js]: are there any standard styling that I can follow?
i.e. `class` and `<div>` or `<p>`, etc

**Comment by paradite** [comment]
@wkurniawan07 updated accordingly.
I used `<p id="status"></p>`
and
`d3.select('#status').text('An error has occured in getting data, please try reloading.');`

**Comment by wkurniawan07** [comment]
One last thing: can you post a screenshot of how it looks like if the loading fails? You can simulate it by changing the json data source to something non-existent.

**Comment by paradite** [comment]
@wkurniawan07 
The Hogeschool van Amsterdam, University of Applied Sciences, or Amsterdam University of Applied Sciences, is the largest institutes for higher professional education in the Netherlands.

**Comment by paradite** [comment]
["Amsterdam University of Applied Sciences","Netherlands"],

**Comment by paradite** [comment]
screen shot 2016-03-15 at 12 45 39 am
@wkurniawan07

**Comment by wkurniawan07** [comment]
I'd say just overwrite the entire `#contentHolder`. There's not much point showing an empty block and the last updated date when an error occurs.
And one last thing, add a line break and this text: "If the problem persists after a few retries, please <a href="contact.html">contact us</a>."

After the change is applied, post another screenshot.

**Comment by paradite** [comment]
screen shot 2016-03-15 at 12 58 17 am
@wkurniawan07

**Comment by paradite** [comment]
screen shot 2016-03-15 at 1 01 27 am

**Comment by wkurniawan07** [comment]
Well done :+1:

**Pull request merged**